easypy/typed_struct.py

```diff
@@ -0,0 +1,366 @@
+from copy import deepcopy
+
+from .exceptions import TException
+from .tokens import MANDATORY
+from .collections import listify, ListCollection, PythonOrderedDict
+from .bunch import Bunch, bunchify
+
+
+class InvalidFieldType(TException):
+    template = 'Bad field type {field_type} - {reason}'
+
+
+class NotFields(TException, AttributeError):
+    template = '{field_names} are not fields of {typed_struct.__name__}'
+
+
+class NotAField(TException, AttributeError):
+    template = '{field_name} is not a field of {typed_struct.__name__}'
+
+
+class NotAKey(TException, KeyError):
+    template = '{field_name} is not a key of {typed_struct.__name__}'
+
+
+class MissingField(TException, TypeError):
+    template = '{typed_struct.__name__} is missing mandatory field {field.name}'
+
+
+class FieldTypeMismatch(TException, TypeError):
+    template = '{typed_struct.__name__}.{field.name} is {expected_type}, not {received_type}'
+
+
+class InvalidFieldValue(TException, ValueError):
+    template = '{typed_struct.__name__}.{field.name} is assigned invalid value {value}'
+
+
+class FieldKeyTypeMismatch(TException, TypeError):
+    template = 'keys of {typed_struct.__name__}.{field.name} are {expected_type}, not {received_type}'
+
+
+class Field(object):
+    """
+    Define a single field in a TypedStruct
+
+        type: The type of the field
+              * Use [<type>] for a typechecked list
+              * Use {<key_type>: <type>} for a typechecked dict
+              * Use {str: <type>} for a typechecked Bunch
+        default: Default value for the field
+        validate: Validator predicate(s) that will be called on the value.
+                  Should return False if it's no OK.
+        convert: Dictionary of converters from source type to the field's type.
```

Comment:
in plumbum the validator is also a converter (returns the normalized value) - couldn't we do that here too?